File: rust/src/bot/digest.rs
//! Engelsystem shift-start digest. Pure stepper + a looping driver.

use chrono::{DateTime, Duration, FixedOffset, Local, NaiveDateTime, NaiveTime, TimeZone, Utc};
use sqlx::SqlitePool;
use std::collections::HashSet;

use super::{notify, Outbox};
use crate::config::AppConfig;
use crate::engelsystem::EngelsystemClient;
use crate::i18n;
use crate::models::now_utc;

pub fn is_within_window(now_local: NaiveTime, start: NaiveTime, end: NaiveTime) -> bool {
    if start == end {
        return true; // always-on
    }
    if start < end {
        return start <= now_local && now_local < end;
    }
    now_local >= start || now_local < end
}

fn parse_iso(s: &str) -> Option<DateTime<FixedOffset>> {
    DateTime::parse_from_rfc3339(s).ok()
}

fn format_shift_announcement(shift: &serde_json::Value, location_name: &str) -> String {
    let local = shift
        .get("starts_at")
        .and_then(|v| v.as_str())
        .and_then(parse_iso)
        .map(|dt| {
            let l: DateTime<Local> = dt.with_timezone(&Local);
            l.format("%H:%M").to_string()
        })
        .unwrap_or_else(|| "??".to_string());

    let mut lines = vec![i18n::digest_next_shift(&local, location_name)];
    let mut has_entries = false;
    if let Some(entries) = shift.get("entries").and_then(|v| v.as_array()) {
        for entry in entries {
            let kind = entry
                .get("type")
                .and_then(|t| t.get("name"))
                .and_then(|n| n.as_str())
                .unwrap_or("?");
            if let Some(users) = entry.get("users").and_then(|u| u.as_array()) {
                for user in users {
                    let name = user.get("name").and_then(|n| n.as_str()).unwrap_or("?");
                    lines.push(format!("- {} [{}]", name, kind));
                    has_entries = true;
                }
            }
        }
    }
    if !has_entries {
        lines.push(i18n::DIGEST_NO_ENTRIES.to_string());
    }
    lines.join("\n")
}

/// One pass of the digest. Returns the number of new shifts announced.
/// `announced` is mutated in place so the caller (the loop) can keep
/// state across passes.
#[allow(clippy::too_many_arguments)]
pub async fn digest_once(
    outbox: &mut Outbox,
    pool: &SqlitePool,
    config: &AppConfig,
    client: &EngelsystemClient,
    announced: &mut HashSet<i64>,
    now: NaiveDateTime,
) -> anyhow::Result<usize> {
    let lookahead = Duration::minutes(config.shift_digest.lookahead_minutes as i64);
    let now_utc = Utc.from_utc_datetime(&now);
    let cutoff = now_utc + lookahead;
    let helfer_orga = config.route_category("Helfer").to_string();
    let mut new_count = 0;

    for (location_name, location_id) in &config.locations {
        let shifts = match client.shifts_at(*location_id).await {
            Ok(s) => s,
            Err(e) => {
                tracing::warn!(
                    "digest: shift fetch for {:?} (id={}) failed: {}",
                    location_name,
                    location_id,
                    e
                );
                continue;
            }
        };
        for shift in shifts {
            let Some(id) = shift.get("id").and_then(|v| v.as_i64()) else {
                continue;
            };
            if announced.contains(&id) {
                continue;
            }
            let Some(start) = shift
                .get("starts_at")
                .and_then(|v| v.as_str())
                .and_then(parse_iso)
            else {
                continue;
            };
            let start_utc = start.with_timezone(&Utc);
            if !(now_utc <= start_utc && start_utc < cutoff) {
                continue;
            }
            let body = format_shift_announcement(&shift, location_name);
            notify::group_msg(outbox, pool, &helfer_orga, &body, None).await?;
            announced.insert(id);
            new_count += 1;
        }
    }
    Ok(new_count)
}

/// Long-running loop. Cancellation drops out cleanly via the await on
/// `tokio::time::sleep`.
#[allow(dead_code)]
pub async fn shift_digest_loop(
    pool: SqlitePool,
    config: AppConfig,
    client: EngelsystemClient,
) -> anyhow::Result<()> {
    let sd = &config.shift_digest;
    if !sd.enabled {
        tracing::info!("shift digest disabled in config");
        return Ok(());
    }
    tracing::info!(
        "shift digest active: window {}–{} local, every {} min, lookahead {} min",
        sd.window_start,
        sd.window_end,
        sd.check_interval_minutes,
        sd.lookahead_minutes
    );
    let interval = std::time::Duration::from_secs(sd.check_interval_minutes * 60);
    let mut announced: HashSet<i64> = HashSet::new();
    loop {
        let now = now_utc();
        let now_local = Utc.from_utc_datetime(&now).with_timezone(&Local).time();
        if is_within_window(now_local, sd.window_start, sd.window_end) {
            let mut outbox = Outbox::new();
            if let Err(e) =
                digest_once(&mut outbox, &pool, &config, &client, &mut announced, now).await
            {
                tracing::warn!("digest pass failed: {}", e);
            }
            // In the loop driver we don't have a Bot to consume the
            // outbox; the teloxide adapter wraps this fn and flushes.
            // Tests construct the outbox themselves and call digest_once.
            // For now, log if a non-empty outbox slipped through.
            if !outbox.actions.is_empty() {
                tracing::trace!("digest produced {} actions", outbox.actions.len());
            }
        }
        tokio::time::sleep(interval).await;
    }
}

File: src/unifestbestellbot/bot/digest.py
"""Engelsystem shift-start digest.

Periodically polls Engelsystem for shifts starting soon and DMs the
orga group that handles `Helfer` requests with the rota of each
upcoming shift. The "starting soon" window and the operational-hours
window come from `config.shift_digest`."""

import asyncio
import logging
from datetime import UTC, datetime, time, timedelta

from aiogram import Bot

from .. import i18n
from ..config import AppConfig
from ..db import session_scope
from ..engelsystem import EngelsystemClient, iter_rota
from ..models import now_utc, to_local
from ..settings import get_settings
from . import notify

log = logging.getLogger(__name__)


def is_within_window(now_local: time, start: time, end: time) -> bool:
    """Whether `now_local` falls inside the configured operational window.

    `start == end` is treated as "always on" so an operator who hasn't
    configured a window doesn't get a permanently-disabled feature."""
    if start == end:
        return True
    if start < end:
        return start <= now_local < end
    # window wraps midnight
    return now_local >= start or now_local < end


def _parse_iso_to_naive_utc(s: str) -> datetime:
    """Engelsystem returns aware ISO timestamps. Bring them into the
    project's naive-UTC convention so they're comparable with `now_utc()`."""
    return datetime.fromisoformat(s).astimezone(UTC).replace(tzinfo=None)


def format_shift_announcement(shift: dict, location_name: str) -> str:
    """Build the DM body for a single upcoming shift. Local time, plus
    the rota with each entry's user and role."""
    start_local = (
        datetime.fromisoformat(shift["starts_at"])
        .astimezone(get_settings().local_tz())
        .strftime("%H:%M")
    )
    lines = [
        i18n.DIGEST_NEXT_SHIFT.format(location=location_name, time=start_local),
    ]
    has_entries = False
    for name, role in iter_rota(shift):
        lines.append(f"- {name} [{role}]")
        has_entries = True
    if not has_entries:
        lines.append(i18n.DIGEST_NO_ENTRIES)
    return "\n".join(lines)


async def digest_once(
    *,
    bot: Bot,
    config: AppConfig,
    client: EngelsystemClient,
    announced: set[int],
    now: datetime | None = None,
) -> int:
    """One pass: for every configured location, fetch shifts and DM the
    Helfer-orga about any shift that starts within the lookahead window
    and hasn't already been announced. Returns the number of new shifts
    announced this pass."""
    now = now if now is not None else now_utc()
    lookahead = timedelta(minutes=config.shift_digest.lookahead_minutes)
    cutoff = now + lookahead
    helfer_orga = config.route_category("Helfer")
    new_count = 0

    for location_name, location_id in config.locations.items():
        try:
            shifts = await client.shifts_at(location_id)
        except Exception:
            log.exception("digest: shift fetch for %r (id=%s) failed",
                          location_name, location_id)
            continue

        for shift in shifts:
            shift_id = shift.get("id")
            if shift_id is None or shift_id in announced:
                continue
            try:
                start = _parse_iso_to_naive_utc(shift["starts_at"])
            except (KeyError, ValueError):
                continue
            if not (now <= start < cutoff):
                continue

            body = format_shift_announcement(shift, location_name)
            with session_scope() as s:
                await notify.group_msg(bot, s, helfer_orga, body)
            announced.add(shift_id)
            new_count += 1
            log.info(
                "digest: announced shift %s at %s starting %s",
                shift_id, location_name, shift["starts_at"],
            )

    return new_count


async def shift_digest_loop(
    *,
    bot: Bot,
    config: AppConfig,
    client: EngelsystemClient,
) -> None:
    """Long-running task. Cancellation drops out cleanly."""
    sd = config.shift_digest
    if not sd.enabled:
        log.info("shift digest disabled in config")
        return

    log.info(
        "shift digest active: window %s–%s local, every %d min, lookahead %d min",
        sd.window_start.strftime("%H:%M"),
        sd.window_end.strftime("%H:%M"),
        sd.check_interval_minutes,
        sd.lookahead_minutes,
    )
    # Dedup of already-announced shift ids is in-memory only: a process
    # restart re-announces any shift still inside the lookahead window. The
    # supervisor (see __main__) makes restarts rare, the window is minutes
    # wide, and a duplicate "shift starting" DM is harmless, so this is left
    # unpersisted deliberately rather than adding a table.
    announced: set[int] = set()
    interval_sec = sd.check_interval_minutes * 60

    try:
        while True:
            now_naive = now_utc()
            now_local = to_local(now_naive).time()
            if is_within_window(now_local, sd.window_start, sd.window_end):
                try:
                    await digest_once(
                        bot=bot, config=config, client=client, announced=announced,
                        now=now_naive,
                    )
                except Exception:
                    log.exception("digest pass failed")
            await asyncio.sleep(interval_sec)
    except asyncio.CancelledError:
        log.info("shift digest loop cancelled")
        raise
